lib/chrontext/src/web.rs

```diff
@@ -0,0 +1,61 @@
+use std::sync::Arc;
+
+use axum::response::Html;
+use axum::Form;
+use axum::{self, Router};
+use axum::{extract::State, routing::get};
+
+use oxigraph::sparql::{results::QueryResultsFormat, QueryResults, QuerySolutionIter};
+use oxrdf::Variable;
+use serde::Deserialize;
+
+use crate::sparql_database::SparqlQueryable;
+
+#[derive(Clone)]
+struct AppState {
+    sparql_engine: Arc<(dyn SparqlQueryable)>,
+}
+
+pub async fn launch_web(sparql_engine: Arc<(dyn SparqlQueryable)>, address: &str) {
+    let state = AppState { sparql_engine };
+
+    let app: Router = Router::new()
+        .route("/", get(|| async { "Hello, World!" }))
+        .route("/query", get(get_query).post(post_query))
+        .with_state(state);
+
+    let listener = tokio::net::TcpListener::bind(address).await.unwrap();
+    axum::serve(listener, app).await.unwrap();
+}
+
+async fn get_query() -> Html<&'static str> {
+    Html(include_str!("web/sparql.html"))
+}
+
+#[derive(Deserialize, Debug)]
+struct SparqlQuery {
+    query: String,
+}
+
+async fn post_query(State(state): State<AppState>, Form(form): Form<SparqlQuery>) -> String {
+    let sparql_engine = state.sparql_engine;
+
+    let query = spargebra::Query::parse(&form.query, None).unwrap();
+
+    let query_result: QueryResults = match sparql_engine.execute(&query).await {
+        Ok(v) => {
+            let variables: Arc<[Variable]> = v.first().unwrap().variables().into();
+            let iter = v
+                .into_iter()
+                .map(|qs| Ok(qs.values().iter().map(|t| t.clone()).collect()));
+            let qsi = QuerySolutionIter::new(variables, iter);
+            qsi.into()
```

Comment:
This is pretty ugly,  but [SparqlQueryable outputs these rows, instead of a real query result](https://github.com/DataTreehouse/chrontext/blob/0ad95cdd94aef7b3f02c1f92e71169c49bc461b5/lib/chrontext/src/sparql_database.rs#L20)

Maybe there should be a `query` function for `SparqlQueryable` which returns a `QueryResults` that makes this all go away?

Reply:
I guess we will run into a similar problem when changing to https://github.com/DataTreehouse/maplib/blob/main/lib/representation/src/polars_to_rdf.rs 
Feel free to refactor if you find a prettier way of doing things :-)